### PR TITLE
Protocol Negotiation

### PR DESCRIPTION
**Pull request by yassernasc**

### CODE REVIEW

**Review by kasperisager — CHANGES_REQUESTED** [review]
Solid start! Now we need to handle all the subtleties:

1. On upgrade, both server and client connections should detach all socket listeners as the socket is now owned by the caller. Any client request tied to the socket should also be closed.
2. If no `upgrade` handler is installed, the socket should be immediately destroyed.
3. Any outstanding body data after the HTTP header should be passed through the `upgrade` event, which will currently always emit an empty buffer.

You can test 3. by passing data to `request.end()` on the client side and by writing data after `handshake` on the server side. This data should be available on the receiving end through the `head` argument. From that point on, the caller is responsible for consuming data from the socket.

**Comment by yassernasc** [comment]
I think some details are missing but a second review would be good for sanity checks,

#### 1, commit: https://github.com/holepunchto/bare-http1/pull/11/commits/b25dc755862c306b1e09b4b9b0e8ce5f06908c43

Idk if all events needed are being detach, especially those from the `client connection`, if I detach the `close` event, it breaks the `close connection if missing upgrade handler` unit test (from our topic 3).

I also saw in some logs that the `end` event is not detached at node.js.

I am having trouble creating unit tests for this issue: `Any client request tied to the socket should also be closed.`, that check should be done at both `Client connection` and `Server connection`?

#### 2, commit:  https://github.com/holepunchto/bare-http1/pull/11/commits/9b901c7e6fcdb9a924f761fe6922219a4b5162b9

Added new logic only for the `request upgrade` because this node.js doc: https://nodejs.org/api/http.html#event-upgrade_1 ("Listening to this event is optional")

#### 3, commit: https://github.com/holepunchto/bare-http1/pull/11/commits/153e906f54a12cf71fd05e904de400908cc33e3f

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: lib/client-connection.js
Comment (kasperisager) [reply lib/client-connection.js]: I've addressed this in 07ce84e.

**Comment by kasperisager** [comment]
> Idk if all events needed are being detach, especially those from the client connection, if I detach the `close` event, it breaks the close connection if missing upgrade handler unit test (from our topic 2), with the error Socket hungs up.

_All_ events should be detached as, after emitting `upgrade`, the socket is entirely owned by the caller. You can find the equivalent code in Node.js here:

- https://github.com/nodejs/node/blob/1b965270a9c273d4cf70e8808e9d28b9ada7844f/lib/_http_server.js#L930-L935
- https://github.com/nodejs/node/blob/1b965270a9c273d4cf70e8808e9d28b9ada7844f/lib/_http_client.js#L558-L560
- https://github.com/nodejs/node/blob/1b965270a9c273d4cf70e8808e9d28b9ada7844f/lib/_http_client.js#L576-L577

The issue with `close` not being emitted was due to the request not being destroyed on `upgrade`, which I've addressed in 07ce84e.

> I am having trouble creating unit tests for this issue: Any client request tied to the socket should also be closed., that check should be done at both Client connection and Server connection?

You can test that on the client side by making sure that `req.on('close')` is emitted. On the server side, the `req` will have already been closed by the time the `upgrade` event is emitted so that's not trivial to test; let's skip that for now.

**Comment by yassernasc** [comment]
First, thanks for the detailed review and related links.

Maybe silly thing but I was wondering about the 'close' event of `req` parameter at server-side upgrade event.

[Node.js docs](https://nodejs.org/api/http.html#event-upgrade_1) only speaks about the `data` event: "After this event is emitted, the request's socket will not have a 'data' event listener (..)"

In the [source code](https://github.com/nodejs/node/blob/1b965270a9c273d4cf70e8808e9d28b9ada7844f/lib/_http_server.js#L932C5-L932C51) we see many events being detached. And by running the following sample code only the `req-side` is logged.

```js
const server = http.createServer()
server.on('upgrade', (req, socket, head) => {
  req.on('close', () => console.log('[server] request closed'))

  socket.end(
    'HTTP/1.1 101 Web Socket Protocol Handshake\r\n' +
      'Upgrade: weird-protocol\r\n' +
      'Connection: Upgrade\r\n' +
      '\r\n'
  )
})

server.listen(1337, () => {
  const req = http.request({
    port: 1337,
    headers: { Connection: 'Upgrade', Upgrade: 'weird-protocol' }
  }).end()

  req.on('upgrade', (res, socket, head) => {
    req.on('close', () => console.log('[req] request closed'))

    socket.end()
  })
})
```

Continuing..

At our current implementation, we don't bind a callback to be detached and yet, the event 'close' is triggered motivated by the `req.destroy()` line at the `_onupgrade` fn.

Later, I started to hope that the current behavior was the intended and the `socket` parameter having a `close` event-listener wasn't a wrong behavior, which makes sense following the hint:

> On the server side, the req will have already been closed by the time the upgrade event is emitted

So I updated the test:

```js
req.on('close', () => t.fail('request close event listener should be detached'))
```

to:

```js
req.on('close', () => up.pass('request closed after server upgrade event'))
```

After running the tests many times and now passing by CI, the `req` closure after the `upgrade event` seems to be deterministic.

**Comment by yassernasc** [comment]
Sorry about the blog post 😅, tell me please if I made a wrong assumption and/or if there's more work to do on this Pull request.

**Comment by kasperisager** [comment]
All good, I don't mind reading! 😂 Your assumptions are correct, the only thing left now is to land this! 👏

**Pull request merged**
